### MSC_DA_CA2

Dataset: EU Travel Survey on demand for innovative transport systems

https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/P82V9X
Description 	
The dataset comes from an EU-wide survey carried out in 2014 with the objective of gathering a number of transport and mobility indicators on transport user preferences with emphasis on the potential of emerging transport technologies and the acceptability of various transport policy measures. The CAWI (Computer Aided Web Interview) survey covered all 28 Member States of the European Union with the same questionnaire translated in the local languages. Samples of 1000 individuals in each country reflected the composition of adult population (from 16 years on) in terms of gender, age class, employment status, education level and living region. The report describing the survey can be accessed at: http://publications.jrc.ec.europa.eu/repository/handle/JRC96151

Keyword 	transport, mobility, survey, European Union

@data{DVN/P82V9X_2016,
author = {Christidis, Panayotis},
publisher = {Harvard Dataverse},
title = {{EU Travel Survey on demand for innovative transport systems}},
UNF = {UNF:6:dVj92hLsh1poq2//aVgVGA==},
year = {2016},
version = {V1},
doi = {10.7910/DVN/P82V9X},
url = {https://doi.org/10.7910/DVN/P82V9X}
}

In [135]:
#Importing the essential libraries, which are Pandas, Matplotlib, Numpy, and Seaborn.
import pandas as pd
from textblob import TextBlob

import warnings
warnings.filterwarnings('ignore')

# Set to display all columns
pd.set_option('display.max_columns', None)

When attempting to read a CSV file that is 'not encoded in UTF-8', which is the default encoding assumed by pandas.read_csv(), a UnicodeDecodeError was encountered. To read a file encoded differently, the alternative was to use the technique 'encoding='ISO-8859-1'.

In [136]:
eu_transp = pd.read_csv('eu_survey_transport_systems.csv', encoding='ISO-8859-1')
eu_transp.head(2)

,ID,Country,Gender,Age,Education,Region,Profession,Work_status,Household_members,Income_level,Location_of_resudence,Centre_or_suburbs,Public_transport_service,Car_driving_license,Number_vehicles_in_household,Considering_electric_or_hybrid_vehicle_next_purchase,Know_what_car_sharing_is,Would_subsribe_car_sharing_if_available,Most_frequent_trip_Walk,Most_frequent_trip_Bicycle,Most_frequent_trip_Car_as_Driver,Most_frequent_trip_Car_as_Passenger,Most_frequent_trip_Train,Most_frequent_trip_Underground_or_light_train,Most_frequent_trip_Tram,Most_frequent_trip_Bus,Most_frequent_trip_Motorcycle_or_moped,Destination_most_frequent_trip,Frequency_most_frequent_trip,Problem_most_frequent_trip_Congestion,Problem_most_frequent_trip_Parking,Problem_most_frequent_trip_Lack_of_bicycle_lanes,Problem_most_frequent_trip_Infrequent_public_transport,Problem_most_frequent_trip_Lack_of_public_transport_coverage,Problem_most_frequent_trip_none,Transfers_between_modes_during_frequent_trip,Frequent_trip_duration_in_minutes,Frequent_trip_distance,Concern_environmental_impacts,Preference_tolls_or_traffic_limitation
0,1,Belgium,Female,49,Upper secondary (high school or similar);,Prov. Oost-Vlaanderen,housewife,Not Employed,two,lower middle,Metropolitan area of a big city with more than...,in the suburbs,Well served by public transport,Yes,1,Maybe yes maybe not,Yes,"Maybe yes, maybe not. I would need to test the...",No,No,Yes,No,No,No,No,No,No,It is outside an urban area,Make this trip every day/ every working day of...,Yes,Yes,No,No,No,No,,20,3-5 KM,5,No preferences
1,2,France,Male,26,"Tertiary and higher (University degree, PhD or...",Pays de la Loire,unemployed,Not Employed,four,low,Small or medium town (less than 250.000 inhabi...,in the centre of the city,Difficult to reach with public transport,Yes,3,Probably not,Yes,"Maybe yes, maybe not. I would need to test the...",No,No,Yes,No,No,No,No,No,No,"In an urban area, different from where I live",Make this trip 2-3 days per week,No,Yes,No,Yes,Yes,No,,20,11-20 KM,8,Probably more acceptable to limit road traffic


In [137]:
eu_transp.shape

(26605, 40)

### Data Preprocessing:

In [138]:
eu_transp.isnull().values.any()

False

In [139]:
# Print the unique country values
eu_transp = pd.DataFrame(eu_transp)
unique_countries = eu_transp['Country'].unique()
print(unique_countries)

['Belgium' 'France' 'Czech Republic' 'Sweden' 'Poland' 'Hungary' 'Germany'
 'Spain' 'Denmark' 'Italy' 'Finland' 'Netherlands' 'Romania' 'Portugal'
 'Lithuania' 'Ireland' 'Greece' 'Great Britain' 'Estonia' 'Slovenia'
 'Latvia' 'Croatia' 'Austria' 'Bulgaria' 'Slovakia' 'Malta' 'Louxembourg'
 'Cyprus']


#### Filter the Dataset for Ireland and Other Countries

In [140]:
# List of Specific Countries
countries = ['Ireland', 'Denmark', 'Belgium', 'Norway', 'Finland', 'Portugal']
filtered_df = eu_transp[eu_transp['Country'].isin(countries)]

# Select Relevant Columns for Sentiment Analysis
# Lowercasing the text
for col in ['Considering_electric_or_hybrid_vehicle_next_purchase', 
            'Would_subsribe_car_sharing_if_available', 
            'Preference_tolls_or_traffic_limitation']:
    filtered_df[col] = filtered_df[col].str.lower()

# Keep the 'Country' column along with the sentiment columns
sentiment_df = filtered_df[['Country'] + sentiment_columns]
sentiment_df.head()

,Country,Considering_electric_or_hybrid_vehicle_next_purchase,Would_subsribe_car_sharing_if_available,Preference_tolls_or_traffic_limitation
0,Belgium,maybe yes maybe not,"maybe yes, maybe not. i would need to test the...",no preferences
5,Belgium,don't know/no answer,don't know / no answer,no preferences
9,Belgium,maybe yes maybe not,"no, i would not be interested in this service",probably more acceptable to limit road traffic
13,Denmark,certainly not,"no, i would not be interested in this service",no preferences
15,Finland,probably yes,"yes, instead of purchasing a new car",probably more acceptable to pay for less conge...


In [141]:
sentiment_df.shape

(5029, 4)

In [142]:
# Print the unique country values
sentiment_df = pd.DataFrame(sentiment_df)
unique_countries = sentiment_df['Country'].unique()
print(unique_countries)

['Belgium' 'Denmark' 'Finland' 'Portugal' 'Ireland']


## Aggregate and Analyze Sentiments 

Create a Sentiment Analysis Function: Using TextBlob or another sentiment analysis tool, define a sentiment analysis function.

Apply the following function on your text columns: For each sentiment analysis result, add a new column to dataset.

In [143]:
# Define the sentiment analysis function
def get_sentiment(text):
    if pd.isna(text):
        return None
    return TextBlob(text).sentiment.polarity

# List of columns to analyze
columns_to_analyze = [
    'Considering_electric_or_hybrid_vehicle_next_purchase', 
    'Would_subsribe_car_sharing_if_available', 
    'Preference_tolls_or_traffic_limitation'
]
# Apply sentiment analysis to each relevant column
for col in columns_to_analyze:
    sentiment_col_name = col + '_sentiment'
    sentiment_df[sentiment_col_name] = sentiment_df[col].apply(get_sentiment)

# Display the first few rows of the DataFrame with new sentiment columns
sentiment_df.head()

,Country,Considering_electric_or_hybrid_vehicle_next_purchase,Would_subsribe_car_sharing_if_available,Preference_tolls_or_traffic_limitation,Considering_electric_or_hybrid_vehicle_next_purchase_sentiment,Would_subsribe_car_sharing_if_available_sentiment,Preference_tolls_or_traffic_limitation_sentiment
0,Belgium,maybe yes maybe not,"maybe yes, maybe not. i would need to test the...",no preferences,0.000000,0.000000,0.000000
5,Belgium,don't know/no answer,don't know / no answer,no preferences,0.000000,0.000000,0.000000
9,Belgium,maybe yes maybe not,"no, i would not be interested in this service",probably more acceptable to limit road traffic,0.000000,0.250000,0.500000
13,Denmark,certainly not,"no, i would not be interested in this service",no preferences,-0.107143,0.250000,0.000000
15,Finland,probably yes,"yes, instead of purchasing a new car",probably more acceptable to pay for less conge...,0.000000,0.136364,0.166667


#### Analyze and Compare Sentiments
This could involve grouping by country, taking averages, plotting results.

This aggregation will display the average sentiment in each selected country for each component of transportation (electric/hybrid vehicle consideration, car sharing subscription, and tolls/traffic constraints). The analysis should then be refined to meet specific research objectives.

In [144]:
# Aggregate the average sentiment by country
average_sentiments = sentiment_df.groupby('Country').agg({
    'Considering_electric_or_hybrid_vehicle_next_purchase_sentiment': 'mean',
    'Would_subsribe_car_sharing_if_available_sentiment': 'mean',
    'Preference_tolls_or_traffic_limitation_sentiment': 'mean'
})

# Display the average sentiments by country
print(average_sentiments)

          Considering_electric_or_hybrid_vehicle_next_purchase_sentiment  \
Country                                                                    
Belgium                                           -0.006000                
Denmark                                           -0.010184                
Finland                                            0.002452                
Ireland                                            0.018857                
Portugal                                           0.026099                

          Would_subsribe_car_sharing_if_available_sentiment  \
Country                                                       
Belgium                                            0.125795   
Denmark                                            0.153704   
Finland                                            0.112759   
Ireland                                            0.118905   
Portugal                                           0.054568   

          Preference_toll

The sentiment analysis results represent the average sentiment scores for each of the three transportation aspects (electric/hybrid vehicle consideration, car sharing subscription, and tolls/traffic limitation preferences) in each of the selected countries (Belgium, Denmark, Finland, Ireland, and Portugal).

- Sentiment Scores: Sentiment scores normally vary between -1.0 and 1.0. A score around -1.0 indicates a strongly negative sentiment, a score near 1.0 shows a strongly positive feeling, and a score near 0 suggests a neutral sentiment or a lack of distinct positive/negative language.

- Considering Electric or Hybrid Vehicle Next Purchase Sentiment: The scores here show the general sentiment towards purchasing an electric or hybrid vehicle in the future. Belgium, for example, has an average score of roughly -0.006, indicating a slightly negative or neutral mood, but Portugal has an average score of approximately 0.026, indicating a slightly positive sentiment.
- Would Subscribe Car Sharing if Available Sentiment: These scores show attitudes towards using automobile sharing services if they are available. Higher positive scores (such as Denmark's 0.1537) imply more positive attitudes towards car sharing services.

- Preference Tolls or Traffic Limitation Sentiment: These scores represent people's feelings towards tolls or traffic restrictions. Positive results suggest a favourable attitude towards these measures. For example, Finland's score of roughly 0.2806 indicates a generally good attitude towards tolls or traffic restrictions.

Conclusion: The sentiment scores provide insight into the general sentiments or ideas held by people in these nations about various transportation-related topics.
A positive score shows a proclivity for positive opinions, a negative score suggests a proclivity for negative opinions, and scores near zero indicate a proclivity for neutral thoughts or a combination of positive and negative attitudes.
These findings can be used to assess public perception or adoption of various transport policies or technologies across countries. However, it is crucial to note that the manner in which emotions are represented can vary substantially depending on cultural and language situations. These ratings should be viewed with these potential variances in mind.

### Machine Learning Classification 
Let's us use a machine learning model to perform a categorization problem.

Classifying the sentiment of the responses into categories such as "Positive," "Negative," and "Neutral." This will entail converting textual input to numerical format and then training a classifier.

# Additional Notes:
Labeling Data: If your dataset doesn't already have sentiment labels, you'll need to label your data manually or use a pre-trained sentiment analysis model to generate these labels.
Choosing a Model: While Naive Bayes is a good starting point, consider experimenting with different models (like Logistic Regression, SVM, or even neural networks) depending on your dataset and accuracy requirements.
Parameter Tuning: Experiment with different parameters in CountVectorizer and your model to optimize performance.
This guide provides a basic framework for text classification. You may need to adjust the steps based on the specifics of your dataset and analysis goals.

In [145]:
# Importing the required libraries
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import classification_report, accuracy_score
from nltk.corpus import stopwords
import nltk

#### Combine and Clean Text Dataset

Let's analyze the sentiment of combined text from multiple columns

In [157]:
# List of Specific Countries
countries = ['Ireland', 'Denmark', 'Belgium', 'Norway', 'Finland', 'Portugal']
transp_sentiment_df = eu_transp[eu_transp['Country'].isin(countries)]

# Combine and clean the text columns:
transp_sentiment_df['combined_text'] = transp_sentiment_df[['Considering_electric_or_hybrid_vehicle_next_purchase', 
                                                                             'Would_subsribe_car_sharing_if_available', 
                                                                             'Preference_tolls_or_traffic_limitation']].agg(' '.join, axis=1)
transp_sentiment_df['combined_text'] = transp_sentiment_df['combined_text'].str.lower()

# Select only the columns you need, such as 'Country' and 'combined_text':
transp_systems_df = transp_sentiment_df[['Country', 'combined_text']]
transp_systems_df.head()

,Country,combined_text
0,Belgium,"maybe yes maybe not maybe yes, maybe not. i wo..."
5,Belgium,don't know/no answer don't know / no answer no...
9,Belgium,"maybe yes maybe not no, i would not be interes..."
13,Denmark,"certainly not no, i would not be interested in..."
15,Finland,"probably yes yes, instead of purchasing a new ..."


In [147]:
transp_systems_df.shape

(5029, 2)

#### Label Creation

In [149]:
# Using TextBlob for automatic labeling
def label_sentiment(text):
    sentiment = TextBlob(text).sentiment.polarity
    if sentiment > 0:
        return 'Positive'
    elif sentiment < 0:
        return 'Negative'
    else:
        return 'Neutral'

transp_systems_df['sentiment_label'] = transp_systems_df['combined_text'].apply(label_sentiment)
transp_systems_df.head(2)

,Country,combined_text,sentiment_label
0,Belgium,"maybe yes maybe not maybe yes, maybe not. i wo...",Neutral
5,Belgium,don't know/no answer don't know / no answer no...,Neutral


In [150]:
# Download and Set Up NLTK Stopwords
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\rosil\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

#### Feature Extraction
Transform text data into a numerical format suitable for machine learning (e.g., using CountVectorizer or TfidfVectorizer).
Create Bag of Words

In [151]:
# Convert the text data into a numerical format using CountVectorizer:
vectorizer = CountVectorizer(stop_words=stopwords.words('english'), max_features=1000)
X = vectorizer.fit_transform(transp_systems_df['combined_text'])

#For demonstration, let's 
y = transp_systems_df['sentiment_label']  

#### Split the dataset into training and testing sets

In [152]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

#### Train a Machine Learning Model

##### Using a Naive Bayes classifier, which is commonly used for text classification:

In [153]:
model = MultinomialNB()
model.fit(X_train, y_train)

MultinomialNB()

#### Model Evaluation
Make Predictions and Evaluate the Model

In [155]:
#Evaluate the model on the test set
y_pred = model.predict(X_test)

print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))

Accuracy: 0.9821073558648111

Classification Report:
               precision    recall  f1-score   support

    Negative       0.48      1.00      0.65        24
     Neutral       1.00      1.00      1.00       272
    Positive       1.00      0.98      0.99      1213

    accuracy                           0.98      1509
   macro avg       0.83      0.99      0.88      1509
weighted avg       0.99      0.98      0.99      1509



- Accuracy (0.9821 or 98.21%): The percentage of total right predictions (both positive and negative) made out of all forecasts made. An accuracy of 98.21% implies that the model accurately identified the sentiment of the text on your test dataset 98.21% of the time.
- The classification report includes additional specific metrics for each class (Negative, Neutral, and Positive):

The precision is defined as the proportion of accurately predicted positive observations to total expected positives. High precision is associated with a low false-positive rate.
- For 'Negative,' 0.48 (48%) indicates that when the model predicts a negative event, it is true 48% of the time.
-For 'Neutral' and 'Positive,' 1.00 (100%) indicates that the model predicts these classifications quite accurately.

Recall (Sensitivity): The proportion of accurately predicted positive observations to the total number of observations in the class.
- For 'Negative,' a value of 1.00 (100%) indicates that the model correctly detected all genuine negative cases.
- The value 0.98 (98%) for 'Positive' shows that the model correctly detected 98% of all actual positive events.
F1-Score: The weighted average of Precision and Recall is the F1-score. It considers both false positives and false negatives. It is a fantastic technique to demonstrate that a class has a good mix of precision and recall.
- For 'Negative': 0.65 (65%) shows a moderate balance between precision and recall for this class.
- Scores for 'Neutral' and 'Positive' are near to 1.00 (100%), suggesting good precision and recall balance.

Support: The number of instances of the specified class in the supplied dataset. In your test set, for example, there are 24 instances of the 'Negative' class.

Macro Average and Weighted Average:
- Macro Avg: The mean precision, recall, and F1-score across all classes. The macro average does not account for label imbalance. The macro average is high in this case (0.83 for precision, 0.99 for recall, and 0.88 for F1-score), but it is skewed by the excellent performance in the 'Neutral' and 'Positive' classes.
- Weighted Avg: The weighted average of precision, recall, and F1-score between classes. When there is a class imbalance, such as in your dataset, this is more useful. The weighted averages are quite high (0.99 for precision, 0.98 for recall, and 0.99 for F1-score), demonstrating good performance across all classes, with a preference for the more frequent classes.

Conclusion: The model does an excellent job of distinguishing 'Neutral' and 'Positive' moods.
However, the model appears to struggle with precision in the 'Negative' class (48%) but has perfect recall, implying that it labels too many instances as negative (high false positives) but does not miss the true negative occurrences.

The high accuracy indicates generally good performance, but the lesser precision for 'Negative' indicates that the model may be over-predicting this class. This could be an area for further research and development, especially in datasets with imbalanced classes.

In [158]:
# Example response from the dataset
text_to_analyze = "maybe yes maybe not"

# Transforming the text
transformed_text = cv.transform([text_to_analyze])

# Predicting probabilities
predicted_probabilities = model.predict_proba(transformed_text)

# Assuming the second class represents a specific sentiment, e.g., 'Positive'
probability_of_second_class = predicted_probabilities[0][1]

print(f"Probability of being 'Positive': {probability_of_second_class}")


NameError: name 'cv' is not defined

### Feature Generation using TF-IDF